src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -216,25 +217,49 @@ private void updateGridMode() {
     sendChannelFocus();
   }
 
+  // When turning a knob that sometimes controls Hue, sometimes Brightness,
+  // sometimes Saturation, look up the current mode of the surface and return
+  // the appropriate subparameter that this knob currently controls.
+  private CompoundParameter getColorSubparam(ColorParameter colorParameter) {
+    if (this.shiftOn) {
+      if (this.bankLeftOn || this.bankRightOn) {
+        return colorParameter.hue;
+      } else {
+        return colorParameter.saturation;
+      }
+    } else if (this.bankLeftOn) {
+      return colorParameter.saturation;
+    } else if (this.bankRightOn) {
+      return colorParameter.brightness;
+    } else {
+      return colorParameter.hue;
+    }
+  }
+
+  private void updateBankLeftRightLights() {
+    sendNoteOn(0, BANK_LEFT, this.bankLeftOn ? LED_ON : LED_OFF);
+    sendNoteOn(0, BANK_RIGHT, this.bankRightOn ? LED_ON : LED_OFF);
+  }
+
   private class DeviceListener implements LXParameterListener {
 
     private LXDeviceComponent device = null;
     private LXEffect effect = null;
     private LXPattern pattern = null;
     private LXBus channel = null;
 
-    private final LXListenableNormalizedParameter[] knobs =
-      new LXListenableNormalizedParameter[DEVICE_KNOB_NUM];
+    private final LXListenableParameter[] knobs =
+            new LXListenableParameter[DEVICE_KNOB_NUM];
 
     DeviceListener() {
-      for (int i = 0; i < this.knobs.length; ++i) {
-        this.knobs[i] = null;
-      }
+      Arrays.fill(this.knobs, null);
```

Comment:
Oh my god how did I never know about that helper. Thanks for this!

Reply:
IntelliJ suggested it; all I did was make a single click.